pkg/controller/job/job_controller_test.go

```diff
@@ -2879,6 +2880,110 @@ func TestSingleJobFailedCondition(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Unexpected error when adding job to indexer %v", err)
+	}
+	// 1st reconcile should create a new pod
+	err = jm.syncJob(ctx, testutil.GetKey(job1, t))
+	if err != nil {
+		t.Fatalf("Unexpected error when syncing jobs %v", err)
+	}
+
+	podIndexer := informer.Core().V1().Pods().Informer().GetIndexer()
+	podList, err := clientSet.Tracker().List(
+		schema.GroupVersionResource{Version: "v1", Resource: "pods"},
+		schema.GroupVersionKind{Version: "v1", Kind: "Pod"},
+		"default")
+	if err != nil {
+		t.Fatalf("Unexpected error when fetching pods %v", err)
+	}
+	// manually adding the just-created pod from fake clientset memory to informer cache because informer is not started.
+	// we are updating the pod status to succeeded which should update the job status to succeeded and remove the finalizer of the pod.
+	justCreatedPod := podList.(*v1.PodList).Items[0]
+	fmt.Printf("pod is %v\n", podList.(*v1.PodList).Items[0])
+	justCreatedPod.Status.Phase = v1.PodSucceeded
+	err = podIndexer.Add(&justCreatedPod)
+	if err != nil {
+		t.Fatalf("Unexpected error when adding pod to indexer %v", err)
+	}
+	jm.addPod(logger, &justCreatedPod)
+	err = jm.syncJob(ctx, testutil.GetKey(job1, t))
+	if err != nil {
+		t.Fatalf("Unexpected error when syncing jobs %v", err)
+	}
+
+	jobList, err := clientSet.Tracker().List(
+		schema.GroupVersionResource{Group: "batch", Version: "v1", Resource: "jobs"},
+		schema.GroupVersionKind{Group: "batch", Version: "v1", Kind: "Job"},
+		"default")
+	if err != nil {
+		t.Fatalf("Unexpected error when trying to get job from the store: %v", err)
+	}
+	updatedJob := jobList.(*batch.JobList).Items[0]
+	if !util.IsJobSucceeded(&updatedJob) {
+		t.Fatalf("job status is not succeeded: %v", updatedJob)
+	}
+
+	// add the updated pod from the fake clientset memory to informer cache because informer is not started.
+	podList, err = clientSet.Tracker().List(
+		schema.GroupVersionResource{Version: "v1", Resource: "pods"},
+		schema.GroupVersionKind{Version: "v1", Kind: "Pod"},
+		"default")
+	if err != nil {
+		t.Fatalf("Unexpected error when fetching pods %v", err)
+	}
+	fmt.Printf("pod is %v\n", podList.(*v1.PodList).Items[0])
+	updatedPod := podList.(*v1.PodList).Items[0]
+	updatedPod.Status.Phase = v1.PodSucceeded
+	err = podIndexer.Add(&updatedPod)
+	if err != nil {
+		t.Fatalf("Unexpected error when adding pod to indexer %v", err)
+	}
+
+	// removing the just created pod from fake clientset memory inorder for the sync job to succeed if creating a new pod because of bug
+	// but the pod will remain inside informer cache
+	err = clientSet.Tracker().Delete(
+		schema.GroupVersionResource{Version: "v1", Resource: "pods"},
+		"default", "")
+	if err != nil {
+		t.Fatalf("Unexpected error when deleting pod to indexer %v", err)
+	}
+
+	err = jm.syncJob(ctx, testutil.GetKey(job1, t))
+	if err != nil {
+		t.Fatalf("Unexpected error when syncing jobs %v", err)
+	}
+	time.Sleep(time.Second)
```

Comment:
IIUC this is need to give time to the Job controller to create the pod when there is a bug. Can you check if 100ms it enough to repro this reliably?